src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -459,6 +471,37 @@ private static string GenerateUpgradeReport(string? projectName, string backupPa
 
                     if (!cancellationToken.IsCancellationRequested)
                     {
+                        if (userAction?.SourceMappingSourceName != null && addedPackages != null)
+                        {
+                            Dictionary<string, IReadOnlyList<string>> patternsReadOnly = existingPackageSourceMappingSourceItems
+                                .ToDictionary(pair => pair.Key, pair => (IReadOnlyList<string>)(pair.Patterns.Select(p => p.Pattern).ToList()));
+
+                            PackageSourceMapping packageSourceMapping = new(patternsReadOnly);
+
+                            // Expand all patterns/globs so we can later check if this package ID was already mapped.
+                            var addedPackageIds = addedPackages.Select(action => action.Item1);
+                            var addedPackageIdsWithoutExistingMappings = new List<string>();
+                            foreach (string? addedPackageId in addedPackageIds)
+                            {
+                                var configuredSource = packageSourceMapping.GetConfiguredPackageSources(addedPackageId);
```

Comment:
This is where the heavy lifting of determining if an existing mapping (pattern or glob) covers the package being added. Only the delta of packages being brought in that were not previously mapped will be mapped.

Possibly this logic could eventually move into the `PackageSourceMappingProvider`. I think it's safe to put here for preview 1 and I can move it in a later iteration if that's a valuable change.